presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java

```diff
@@ -702,6 +712,25 @@ private void generatePositionNotDistinctFromRowWithPageMethod(
                     .cast(Block.class);
             BytecodeExpression rightBlock = page.invoke("getBlock", Block.class, rightChannels.getElement(index));
             Type type = joinChannelTypes.get(index);
+            // This is a huge hack to improve performance for known types with Slice as javaType
+            // Since null is not distinct from null in typeEquals for these types
+            if (type.getJavaType().equals(Slice.class)) {
```

Comment:
This will be another place that would need to be updated when a new type is added. Moreover, it is impossible to update this code for types loaded from plugins.

What if we had a method in `Type` class: `Type#isNotDistinct(...)` instead? It would default to:
```
equals(...)
```

@martint ?

Reply:
The plan is to allow any scalar function optionally provide an additional implementation using Block+position calling  convention. At that point, we'll fully deprecate Type.equalTo (and friends).

Reply:
Makes sense. Thanks for information.

> At that point, we'll fully deprecate Type.equalTo (and friends).

I wonder how easy it would be to remove those. I once tried to remove all static usages of `Type` instances, but it was rather hard.

Reply:
Thank you for this insight.